rust/lance/src/dataset/fragment/write.rs

```diff
@@ -62,11 +61,57 @@ impl<'a> FragmentCreateBuilder<'a> {
     /// Write a fragment.
     pub async fn write(
         &self,
-        reader: impl RecordBatchReader + Send + 'static,
+        source: impl StreamingWriteSource,
         id: Option<u64>,
     ) -> Result<Fragment> {
-        let (stream, schema) = self.get_stream_and_schema(Box::new(reader)).await?;
-        self.write_impl(stream, schema, id).await
+        let id = id.unwrap_or_default();
```

Comment:
Please separate this back out into `write()` and `write_impl`. Each concrete type passed into `write()` will generate a new set of code for it. We have them dispatch into `write_impl` to minimize that size of the code that is duplicated.

Reply:
Because the new `StreamingWriteSource` has to implement `Self: Sized`, I converted this impl into (stream, schema) which is a set of concrete types before entering `_impl` methods.